lib/radiator/media/audio_file_upload.ex

```diff
@@ -58,30 +64,74 @@ defmodule Radiator.Media.AudioFileUpload do
     {:ok, %File.Stat{size: size}} = File.lstat(path)
     mime_type = MIME.from_path(path) |> fix_mime_type()
 
+    attrs = %{
+      "title" => filename,
+      "file" => upload,
+      "mime_type" => mime_type,
+      "byte_length" => size,
+      # some rough approximated default
+      "duration" => ceil(size / 96)
+    }
+
+    attrs =
+      case probe_file(path) do
+        {:ok, additional_attrs} ->
+          Logger.info("additional attributes: #{inspect(additional_attrs)}")
+
+          attrs
+          |> Map.merge(additional_attrs)
+
+        _ ->
+          attrs
+      end
+
     fn %{create_audio_file: audio_file} ->
-      AudioFile.changeset(audio_file, %{
-        "title" => filename,
-        "file" => upload,
-        "mime_type" => mime_type,
-        "byte_length" => size
-      })
+      AudioFile.changeset(audio_file, attrs)
     end
   end
 
-  defp add_audio_file_changeset(upload) when is_binary(upload) do
-    mime_type = MIME.from_path(upload) |> fix_mime_type()
+  defp parse_k(string, suffix) when is_binary(string) do
+    case Float.parse(string) do
+      {float, _rest} ->
+        "#{float / 1_000.0}" <> suffix
 
-    # todo: get byte_length _after_ storing
+      _ ->
+        string
+    end
+  end
 
-    fn %{create_audio_file: audio_file} ->
-      audio_file = Repo.preload(audio_file, :audio)
-
-      AudioFile.changeset(audio_file, %{
-        # "title" => filename,
-        "file" => upload,
-        "mime_type" => mime_type
-        # "byte_length" => size
-      })
+  def probe_file(path) do
+    with {:ok, streams} <- FFprobe.streams(path) do
+      stream =
+        streams
+        |> Enum.find(hd(streams), fn m -> m["codec_type"] == "audio" end)
+
+      duration = ((Float.parse(stream["duration"]) |> elem(0)) * 1_000) |> ceil
+
+      audio_format =
```

Comment:
How about a file_meta column of type map where we can store arbitrary key-value-pairs? Easier to work with than comma separated values as API/frontend might want to display these metas in their own way.

Reply:
yeah - a more clearly defined storage of the extracted metadata is probably prudent. This is more intended as a read only value for now and as a finger excersize in terms of what FFprobe is delivering.